File: src/main/scala/com/formious/services/api/AccessTokens.scala
package com.formious.services.api

import io.circe.generic.auto._
import io.circe.syntax.EncoderOps

import com.formious.models.AccessToken
import com.formious.common.Encoders._
import com.formious.common.Decoders._

import org.http4s._
import org.http4s.dsl._
import org.http4s.circe._

object AccessTokens {
  val service = HttpService {
    case GET -> Root =>
      // List all access tokens
      Ok(AccessToken.all.asJson)
    case GET -> Root / "new" =>
      // Generate blank access token
      Ok(AccessToken.empty.asJson)
    case request @ POST -> Root =>
      // Create new access token
      request.as(jsonOf[AccessToken]).flatMap { bodyAccessToken =>
        val AccessToken(_, token, relation, foreign_id, expires, redacted, _) = bodyAccessToken
        val accessToken = AccessToken.create(token, relation, foreign_id, expires, redacted)
        Created(accessToken.asJson)
      }
    case GET -> Root / IntVar(id) =>
      // Show existing access token
      Ok(AccessToken.find(id).asJson)
    case request @ POST -> Root / IntVar(id) =>
      // Update existing access token
      request.as(jsonOf[AccessToken]).flatMap { bodyAccessToken =>
        val AccessToken(_, token, relation, foreign_id, expires, redacted, _) = bodyAccessToken
        AccessToken.update(id, token, relation, foreign_id, expires, redacted)
        NoContent()
      }
    case DELETE -> Root / IntVar(id) =>
      // Delete existing access token
      AccessToken.delete(id)
      NoContent()
  }
}

File: src/com/formious/routes/api/access_tokens.clj
(ns com.formious.routes.api.access-tokens
  (:require [com.formious.common :refer [no-content created ok]]
            [com.formious.db.access-token :as AccessToken]
            [compojure.coercions :refer [as-int]]
            [compojure.core :refer [GET HEAD PATCH POST PUT DELETE defroutes context]]))

(defroutes routes
  (GET "/" [] ; List all access tokens
    (-> (AccessToken/all) (ok)))
  (GET "/new" [] ; Generate blank access token
    (-> (AccessToken/blank) (ok)))
  (POST "/" {body :body} ; Create new access token
    (->> (select-keys body ["token" "relation" "foreign_id" "expires" "redacted"])
         (AccessToken/insert!)
         (created)))
  (GET "/:id" [id :<< as-int] ; Show existing access token
    (-> (AccessToken/find-by-id id) (ok)))
  (POST "/:id" [id :<< as-int :as {body :body}] ; Update existing access token
    (->> (select-keys body ["token" "relation" "foreign_id" "expires" "redacted"])
         (AccessToken/update! id))
    (no-content))
  (DELETE "/:id" [id :<< as-int] ; Delete existing access token
    (AccessToken/delete! id)
    (no-content)))
